### ISSUE TITLE
L5PC gating creates NaNs

### ISSUE THREAD

**Issue by dgedon** [jaxleyverse/jaxley-mech#70]
# L5PC gates return NaN under `jax.jit` in float32

Several rate equations in `jaxley_mech/channels/l5pc.py` add `1e-6` to a voltage offset to
avoid dividing by zero. In float32 that constant is too small to change the offset it is
added to, so it disappears and the division by zero happens anyway.

```
jax 0.10.1 | jaxlib 0.10.1 | jaxley 0.14.0 | jaxley-mech 0.3.1
python 3.13.11 | Linux x86_64 | backend cpu | x64 False
```

## Reproducer

```python
import jax
import jax.numpy as jnp
from jaxley_mech.channels.l5pc import NaTaT

v = jnp.float32(-38.0)
print("eager m_inf:", NaTaT.m_gate(v)[0])           # 0.5947712  (correct)
print("jit   m_inf:", jax.jit(NaTaT.m_gate)(v)[0])  # nan
```

## Cause

Line 100 is undefined when `v + 38 + 1e-6` is zero:

```python
alpha = (0.182 * (v + 38 + 1e-6)) / (1 - save_exp(-(v + 38 + 1e-6) / 6))
```

Neighbouring float32 values near 38 are `3.81e-06` apart, so `1e-6` cannot change 38:

```python
>>> np.float32(38) + np.float32(1e-6) == np.float32(38)
np.True_
```

Whether the `1e-6` survives then depends on the order of the two additions. Python evaluates
`(v + 38) + 1e-6` left to right and gets a correct answer; XLA combines the two constants
into `38 + 1e-6`, which is just `38`, leaving `v + 38`. At `v = -38` that is exactly zero and
both numerator and denominator become zero:

```
eager : v+38+1e-6 = 1.0e-06   1 - save_exp(...) = 1.79e-07   m_inf = 0.5948
jit   : v+38+1e-6 = 0.0e+00   1 - save_exp(...) = 0.00e+00   m_inf = nan
```

Enabling x64 avoids it, so this is float32 rounding rather than a broken install. Note the
result is correct without `jax.jit` and wrong with it.

## Affected lines

Each rate expression tested at the voltage where its denominator is zero, and the four
nearest float32 values either side:

| lines | channel / gate | voltage | result |
| --- | --- | --- | --- |
| 100–101 | `NaTaT.m_gate` | `-38.0` | NaN |
| 110–111 | `NaTaT.h_gate` | `-66.0` | NaN |
| 182–183 | `NaTs2T.m_gate` | `-32.0`, `-31.999998` | NaN |
| 192–193 | `NaTs2T.h_gate` | `-60.0` | NaN |
| 267–268 | `NapEt2.m_gate` | `-38.0` | NaN |
| 277 | `NapEt2.h_gate` (alpha) | `-17.000002` | NaN |
| 278–279 | `NapEt2.h_gate` (beta) | `-64.4` | fine |
| 723 | `CaHVA.m_gate` | `-26.999998` | NaN |
| 975–976 | `H.m_gate` | `-154.9` | NaN |

Line 278 happens to survive; XLA's reordering is not guaranteed either way, so the exact set
of failing lines should not be relied on. Line 724 has no division. The `H` case is at
`-154.9`, outside the physiological range, so it likely does not matter in practice.

## Suggested fix

Two changes per rate, no `1e-6` in the arithmetic. Replace the zero explicitly, and use
`expm1`, which stays accurate for small arguments where `1 - exp(x)` does not. For line 100:

```python
za = jnp.where(v + 38 == 0, 1e-6, v + 38)
alpha = (0.182 * za) / -jnp.expm1(-za / 6)
```

Substituting `1e-6` is enough because `expm1` then returns the limit `0.182 * 6 = 1.092`
directly, so no second branch is needed for the result.

`m_inf` under `jax.jit`, before and after:

```
v       = [-38.0,     -38.000004, -37.999996, -40.0,     -30.0    ]
current = [nan,        0.617520,   0.571608,   0.5125953, 0.8477496]
fixed   = [0.5947713,  0.5947711,  0.5947714,  0.5125953, 0.8477496]
```

The correct value at `v = -38` is `0.5947712`. Away from the undefined point the numbers are
unchanged. Across nine neighbouring float32 voltages the variation in `m_inf` drops from
`4.6e-02` to `1.2e-06`, and the gradient at `v = -38` becomes finite instead of NaN.

Lines 723 and 975 use `exp(x) - 1` rather than `1 - exp(x)`, so there `expm1(x)` replaces the
denominator directly, without the minus sign.

## Why it matters

These rates are evaluated at every compartment and timestep, easily 10^8 times per run, so
reaching one of these voltages is likely. One NaN then spreads through the coupled cable
equations and the whole cell is NaN on the next timestep. In a 1144-compartment L5PC cell,
three of four stimulus protocols lost 50–70% of the recording, with soma and axon failing on
the same timestep, and only 21 of 100,000 parameter samples produced usable output.

**Comment by huangziwei** [comment]
I believe NaN with float32 is more general problem than channel mechanism. see https://github.com/jaxleyverse/jaxley/issues/317, have you tried float64?

**Comment by dgedon** [comment]
Thanks @huangziwei for the quick reply!

For our work we really would like to have the `float32` option stable. We run a lot of simulations (100k parameter samples for this model), and `float64` doubles the memory and makes it slower. Right now this breaks for us: only 21 out of our 100,000 samples came out with usable data. I also think this one is separate from the general float32 NaN topic, since my example is a single function call with no integration, no training and no morphology involved. See the reproducer code in the original issue.

I think the fix might be small though. The problem is that `1e-6` is too small to show up in `float32`: `float32(38) + 1e-6` gives back exactly `38.0`, so the term disappears and you end up with `0/0`. The original `NEURON` file this was ported from does it differently.NaTa_t.mod in BluePyOpt has if (v == -38) { v = v + 0.0001 }, so 1e-4 and only at that one voltage, see https://github.com/BlueBrain/BluePyOpt/blob/master/examples/l5pc/mechanisms/NaTa_t.mod#L62-L64. I don't think we need to copy that exactly (it probably introduces accuracy errors), but a small change might be sufficient.

**Comment by huangziwei** [comment]
@michaeldeistler do you still want to maintain the L5PC channels yourself or should I?

**Comment by dgedon** [comment]
I have a PR almost ready and can post it soon. We could take it as a basis for further discussions if you like

**Comment by huangziwei** [comment]
Thanks. I also need some time to process it. I am not really working on this at the moment, so it may take a bit of time.

**Comment by michaeldeistler** [comment]
I am fine with bumping it to 1e-4 or 1e-5

**Comment by dgedon** [comment]
Thanks @michaeldeistler but I am not sure this solves the issue. When we run with 1e-4 we just shift the location where nans appear. See:
```python
import jax
import jax.numpy as jnp

EPS = 1e-4  

def m_gate(v):  # line 100 of l5pc.py, with 1e-6 replaced by EPS
    alpha = (0.182 * (v + 38 + EPS)) / (1 - jnp.exp(-(v + 38 + EPS) / 6))
    beta = (0.124 * (-v - 38 + EPS)) / (1 - jnp.exp(-(-v - 38 + EPS) / 6))
    return alpha / (alpha + beta)

v = jnp.float32(-38.000099182128906)  # = -float32(38 + 1e-4)
print("eager m_inf:", m_gate(v))      # 0.62664956
print("jit   m_inf:", jax.jit(m_gate)(v))  # nan
```
This is because we have `v + 38 + 0.0001` is now exactly zero at `v = -38.0001` leading to the same nan issue.

NEURON does it differently, since it applies the EPS only at that one voltage, inside an `if`, rather than adding it everywhere, so no new bad voltage is created.
```
if (v == -38) { v = v + 0.0001 }
alpha = 0.182 * (v + 38) / (1 - exp(-(v + 38) / 6))
```

In #71 we instead correct that one voltage directly, so there is no eps to pick and no shifted zero. It also improves precision at the voltages right next to `-38`, which no choice of eps can fix. 

Maybe the NEURON implementation is less change, happy to adapt the PR. Any thoughts @huangziwei @michaeldeistler ?

**Comment by huangziwei** [comment]
I will soon be on vacation and will be away for >3 weeks. I don't want to rush through it, so I will just let @kyralianaka handle this while I am away, as it can also benefit her work! :)

**Comment by michaeldeistler** [comment]
Ah, I see, that makes sense and sounds good to me
